File: frontend/stores/timeEntry.ts
import { defineStore } from 'pinia'

interface Project {
  id: number
  name: string
  client: string | null
  status: string
}

interface TimeEntry {
  id: number
  user_id: number
  project_id: number | null
  project: Project | null
  start_time: string
  end_time: string | null
  description: string | null
  duration: number | null
  duration_formatted: string | null
  is_active: boolean
}

interface TimeEntryState {
  activeEntry: TimeEntry | null
  entries: TimeEntry[]
  projects: Project[]
  loading: boolean
}

export const useTimeEntryStore = defineStore('timeEntry', {
  state: (): TimeEntryState => ({
    activeEntry: null,
    entries: [],
    projects: [],
    loading: false,
  }),

  actions: {
    async fetchActiveEntry() {
      const { apiFetch } = useApi()

      try {
        const response = await apiFetch<any>('/dashboard')
        this.activeEntry = response?.active_entry || null
        return this.activeEntry
      } catch (error) {
        console.error('Error fetching active entry:', error)
        this.activeEntry = null
        return null
      }
    },

    async fetchEntries(params?: { date?: string; week?: boolean }) {
      const { apiFetch } = useApi()
      this.loading = true

      try {
        const queryParams = new URLSearchParams()
        if (params?.date) queryParams.append('date', params.date)
        if (params?.week) queryParams.append('week', 'true')

        const endpoint = `/time-entries${queryParams.toString() ? `?${queryParams}` : ''}`
        const response = await apiFetch<any>(endpoint)

        this.entries = response?.data || []
        return this.entries
      } catch (error) {
        console.error('Error fetching entries:', error)
        this.entries = []
        return []
      } finally {
        this.loading = false
      }
    },

    async fetchProjects() {
      const { apiFetch } = useApi()

      try {
        const response = await apiFetch<Project[]>('/projects')
        this.projects = response
        return this.projects
      } catch (error) {
        console.error('Error fetching projects:', error)
        this.projects = []
        return []
      }
    },

    async startEntry(projectId?: number, description?: string) {
      const { apiFetch } = useApi()

      try {
        const response = await apiFetch<TimeEntry>('/time-entries/start', {
          method: 'POST',
          body: {
            project_id: projectId || null,
            description: description || null,
          },
        })

        this.activeEntry = response
        return response
      } catch (error: any) {
        console.error('Error starting entry:', error)
        throw error
      }
    },

    async stopEntry() {
      const { apiFetch } = useApi()

      try {
        const response = await apiFetch<TimeEntry>('/time-entries/stop', {
          method: 'POST',
        })

        this.activeEntry = null
        await this.fetchEntries()
        return response
      } catch (error) {
        console.error('Error stopping entry:', error)
        throw error
      }
    },

    async createEntry(data: {
      project_id?: number
      start_time: string
      end_time: string
      description?: string
    }) {
      const { apiFetch } = useApi()

      try {
        const response = await apiFetch<TimeEntry>('/time-entries', {
          method: 'POST',
          body: data,
        })

        await this.fetchEntries()
        return response
      } catch (error) {
        console.error('Error creating entry:', error)
        throw error
      }
    },
  },
})

File: flutter/lib/data/models/time_entry.dart
class TimeEntry {
  final int id;
  final int userId;
  final int? projectId;
  final DateTime startTime;
  final DateTime? endTime;
  final String? description;
  final int? duration;
  final double? latitude;
  final double? longitude;
  final double? locationAccuracy;
  final DateTime? locationCapturedAt;
  final bool qrCodeScanned;
  final String? projectName;
  final bool isActive;

  TimeEntry({
    required this.id,
    required this.userId,
    this.projectId,
    required this.startTime,
    this.endTime,
    this.description,
    this.duration,
    this.latitude,
    this.longitude,
    this.locationAccuracy,
    this.locationCapturedAt,
    required this.qrCodeScanned,
    this.projectName,
    required this.isActive,
  });

  factory TimeEntry.fromJson(Map<String, dynamic> json) {
    return TimeEntry(
      id: json['id'],
      userId: json['user_id'],
      projectId: json['project_id'],
      startTime: DateTime.parse(json['start_time']),
      endTime: json['end_time'] != null ? DateTime.parse(json['end_time']) : null,
      description: json['description'],
      duration: json['duration'],
      latitude: json['latitude'] != null ? double.parse(json['latitude'].toString()) : null,
      longitude: json['longitude'] != null ? double.parse(json['longitude'].toString()) : null,
      locationAccuracy: json['location_accuracy'] != null ? double.parse(json['location_accuracy'].toString()) : null,
      locationCapturedAt: json['location_captured_at'] != null ? DateTime.parse(json['location_captured_at']) : null,
      qrCodeScanned: json['qr_code_scanned'] ?? false,
      projectName: json['project']?['name'],
      isActive: json['is_active'] ?? false,
    );
  }

  String get durationFormatted {
    if (duration == null) return '--:--:--';
    final hours = duration! ~/ 3600;
    final minutes = (duration! % 3600) ~/ 60;
    final seconds = duration! % 60;
    return '${hours.toString().padLeft(2, '0')}:${minutes.toString().padLeft(2, '0')}:${seconds.toString().padLeft(2, '0')}';
  }

  bool get hasLocation => latitude != null && longitude != null;
}
